File: Assign_1/src/FixedPointList.java
// author of program:  Ramish
// Date:  1/28/2015
// Description:  This program is a fixed point list that can hold fixed point number and perform CRUD on them

import java.util.ArrayList;
import java.util.Scanner;

public class FixedPointList {
    final private ArrayList<FixedPointNumber> list;

    public void setListQVal(int listQVal) {
        this.listQVal = listQVal;
    }

    public int getListQVal() {
        return listQVal;
    }

    private int listQVal = 12;

    // constructor
    public FixedPointList() {
        this.list = new ArrayList<>();
    }

    // add to list
    public void add(FixedPointNumber x) {
        this.list.add(x);
        System.out.println(x + " was added to the list");
    }

    // remove first instance of x
    public boolean delete(FixedPointNumber x) {
        for (int i = 0; i < this.list.size(); i++) {
            if (this.list.get(i).equals(x)) {
                this.list.remove(i);
                return true;
            }
        }
        return false;
    }

    // print all fixed point numbers in list
    public void print() {
        System.out.println("All fixed point numbers in list:");
        if (this.list.size() == 0) {
            System.out.println(new FixedPointNumber(0, 12));
        } else {
            for (FixedPointNumber number : this.list) {
                System.out.println(number);
            }
        }
    }

    //find index of first instance of x
    public int find(FixedPointNumber x) {
        return this.list.indexOf(x);
    }

    // change q value of specific fixed point number
    public void changeToQ(FixedPointNumber FpNumber, int q) {
        int index = find(FpNumber);
        FixedPointNumber number = this.list.get(index);
        this.list.set(index, number.toQval(q));
    }

    // change q value of all fixed point numbers in list
    public void changeListQ(int q) {
        System.out.println("Current q_value was changed to " + q);
        setListQVal(q);
    }

    // sum all fixed point numbers in list and print result
    public FixedPointNumber sum(int q) {
        FixedPointNumber sum = new FixedPointNumber(0, q);
        for (FixedPointNumber number : this.list) {
            sum = sum.add(number, this.listQVal);
        }
        return sum;
    }

    // end program
    public void end() {
        System.out.println("Program ended");
    }

    // run program
    public void run() {
        System.out.println("Welcome to the Fixed Point List Program");
        while (true) {
            System.out.print("Enter a command: ");
            Scanner in = new Scanner(System.in);
            String operator = in.next();
            if (operator.equalsIgnoreCase("X")) {
                end();
                break;
            } else if (operator.equalsIgnoreCase("A")) {
                add(new FixedPointNumber(in.nextDouble(), getListQVal()));
            } else if (operator.equalsIgnoreCase("P")) print();
            else if (operator.equalsIgnoreCase("Q")) changeListQ(in.nextInt());
            else if (operator.equalsIgnoreCase("D")) {
                FixedPointNumber num = new FixedPointNumber(in.nextDouble(), listQVal);
                if (delete(num)) {
                    System.out.println(num + " was removed from the list");
                } else {
                    System.out.println(num + " was not found in the list");
                }

            } else if (operator.equalsIgnoreCase("S")) {
                System.out.println("the sum is : " + sum(listQVal));
            } else System.out.println("Invalid command");
        }

    }
}

File: Assign_1_Ruby/lib/FixedPointList.rb
require_relative 'FixedPointNumber'

class FixedPointList
  @fixed_point_list = []
  @list_val = 12
  attr_accessor :list_val

  def initialize
    @fixed_point_list = []
    @list_val = 12
  end

  def set_q_val(q_val)
    @list_val = q_val
  end

  attr_reader :list_val

  #add a fixed point number to the list
  def add_fixed_point(fixed_point)
    @fixed_point_list.push(fixed_point)
    puts "#{fixed_point } was added  to the list"
  end

  #remove first instance of a fixed point number in the list
  def remove_fixed_point(fixed_point)
    (0..@fixed_point_list.length - 1).each { |i|
      #@a =@fixed_point_list[i].to_Q_val(@list_val)
      if @fixed_point_list[i].to_Q_val(@list_val).equals(fixed_point)
        @fixed_point_list.delete_at(i)
        puts "#{fixed_point} was removed from the list"
        return
      end
    }
    puts "#{fixed_point} was not found in the list"
  end

  # print all the fixed point numbers in the list
  def print
    if @fixed_point_list.length == 0 ? puts(FixedPointNumber.new(0, 12).to_s) : @fixed_point_list.each { |i| puts i.to_s }
    end
  end

  # change q value of specific fixed point number in the list
  def change_to_q_val(fixed_point, q_val)
    @fixed_point_list[@fixed_point_list.index(fixed_point)] = fixed_point.to_Q_val(q_val)
  end

  #change list q value
  def change_list_q_val(q_val)
    set_q_val(q_val)
    puts "current q value was changed to #{q_val}"
  end

  #sum all the fixed point numbers in the list
  def sum
    summed = FixedPointNumber.new(0, @list_val)
    @fixed_point_list.each do |fixed_point|
      summed = summed.add(fixed_point, @list_val)
    end
    puts " the sum is: #{summed.to_s}"
  end

  # end program
  def end_program
    puts "Normal termination of program"
    exit
  end

  # run program
  def run
    puts "Welcome to the Fixed Point Number Program"
    puts "Please enter a command"

    while true
      puts "Enter command: "
      command, fixed_point = gets.chomp.chomp.split(' ')
      puts(fixed_point.to_f)
      if command.upcase == "X"
        end_program
      elsif command.upcase == "A" then
        add_fixed_point(FixedPointNumber.new(fixed_point.to_f, @list_val.to_i))
      elsif command.upcase == "D" then
        remove_fixed_point(FixedPointNumber.new(fixed_point.to_f, @list_val.to_i))
      elsif command.upcase == "P" then
        print
      elsif command.upcase == "S" then
        sum
      elsif command.upcase == "Q" then
        change_list_q_val(fixed_point.to_i)
      else
        puts "Invalid command"
      end
    end
  end

end

a = FixedPointList.new
a.run
